File: java-lambda/lambda/src/main/java/com/trace0/javalambda/handlers/StoreUser.java
package com.trace0.javalambda.handlers;

import com.amazonaws.services.lambda.runtime.events.APIGatewayProxyRequestEvent;
import com.amazonaws.services.lambda.runtime.events.APIGatewayProxyResponseEvent;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.trace0.javalambda.Response;
import com.trace0.javalambda.model.User;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import software.amazon.awssdk.services.dynamodb.DynamoDbClient;
import software.amazon.awssdk.services.dynamodb.model.AttributeValue;
import software.amazon.awssdk.services.dynamodb.model.PutItemRequest;

import java.time.Instant;
import java.util.Map;
import java.util.UUID;

public class StoreUser {

    private static final Logger logger = LogManager.getLogger(StoreUser.class);
    private static final DynamoDbClient dynamo = DynamoDbClient.create();
    private static final String TABLE_NAME = System.getenv("USERS_TABLE_NAME");
    private static final ObjectMapper mapper = new ObjectMapper();

    public APIGatewayProxyResponseEvent handle(APIGatewayProxyRequestEvent event) {
        if (event.getBody() == null || event.getBody().isBlank()) {
            return Response.json(400, "{\"error\":\"Request body is required\"}");
        }

        Map<?, ?> body;
        try {
            body = mapper.readValue(event.getBody(), Map.class);
        } catch (Exception e) {
            return Response.json(400, "{\"error\":\"Invalid JSON body\"}");
        }

        String name = (String) body.get("name");
        String email = (String) body.get("email");
        if (name == null || email == null) {
            return Response.json(400, "{\"error\":\"name and email are required\"}");
        }

        String userId = "usr_" + System.currentTimeMillis() + "_" + UUID.randomUUID().toString().substring(0, 7);
        User user = new User(userId, name, email, Instant.now().toString());

        logger.info("Storing user with id: {}", userId);

        try {
            dynamo.putItem(PutItemRequest.builder()
                    .tableName(TABLE_NAME)
                    .item(Map.of(
                            "userId", AttributeValue.fromS(user.getUserId()),
                            "name", AttributeValue.fromS(user.getName()),
                            "email", AttributeValue.fromS(user.getEmail()),
                            "createdAt", AttributeValue.fromS(user.getCreatedAt())
                    ))
                    .build());
        } catch (Exception e) {
            logger.error("Failed to store user", e);
            return Response.json(500, "{\"error\":\"Internal server error\"}");
        }

        logger.info("User stored successfully with id: {}", userId);
        try {
            return Response.json(201, mapper.writeValueAsString(user));
        } catch (Exception e) {
            return Response.json(500, "{\"error\":\"Internal server error\"}");
        }
    }
}

File: node-js-express-docker/app/src/handlers/storeUser.ts
import { Request, Response } from 'express';
import { PutCommand } from '@aws-sdk/lib-dynamodb';
import { dynamo, TABLE_NAME } from '../dynamodb';
import logger from '../logger.js';

interface StoreUserBody {
  name: string;
  email: string;
}

export interface User {
  userId: string;
  name: string;
  email: string;
  createdAt: string;
}

function generateUserId(): string {
  return `usr_${Date.now()}_${Math.random().toString(36).substring(2, 9)}`;
}

export async function storeUser(req: Request, res: Response): Promise<void> {
  const { name, email } = req.body as Partial<StoreUserBody>;

  if (!name || !email) {
    res.status(400).json({ error: 'name and email are required' });
    return;
  }

  const userId = generateUserId();
  const user: User = { userId, name, email, createdAt: new Date().toISOString() };

  logger.info(`Storing user with id: ${userId}.`);

  try {
    if (email === 'trigger-error@example.com') {
      throw new Error(`Invalid email address: ${email}.`);
    }

    await dynamo.send(new PutCommand({ TableName: TABLE_NAME, Item: user }));
  } catch (err) {
    const error = err as Error;
    logger.error({ err: error }, 'Failed to store user');
    res.status(500).json({ error: 'Internal server error' });
    return;
  }

  logger.info(`User stored successfully with id: ${userId}`);
  res.status(201).json(user);
}
